### DISCUSSION TITLE
"Serverless" as the only runtime target

### DISCUSSION THREAD

**Discussion by garnerp** [vercel/next.js#10741] [Help] [answered]
Can anyone clarify what "serverless as the only runtime target" means for the roadmap of next.js or organizations that run outside of NOW?

**Comment by Timer** [comment]
Can you please point to what you're referring to? This doesn't sound like an accurate depiction of what we're doing.

**Comment by garnerp** [comment, marked as answer]
Comments in PR's like this:

#10731

**Reply by timneutkens** [reply]
> Can anyone clarify what "serverless as the only runtime target" means for the roadmap of next.js or organizations that run outside of NOW?

Hey @garnerp! 

Currently Next.js has 2 "output targets":

- Server
- Serverless

The main difference between the 2 is that `serverless` would output "single file bundles" in it's initial form. Later on we added another target called `experimental-serverless-trace`, this target is much more optimized version of `serverless` that works more like how `server` would currently, except it's better than `server` as it no longer requires a full `node_modules` to be shipped.

Our efforts right now are to consolidate all targets into one that covers all cases from server to serverless.

So what does this actually mean?

- Server target builds will be **smaller** and execution will be **faster** than it is currently as they won't require `node_modules` to be shipped with the application
- Serverless target builds will still output very small serverless functions. Main difference with the current serverless target is that when you're not using ZEIT (ZEIT already forces `serverless-trace` instead of `serverless`) you get less build failures because of `node_modules`.
- Less divergence in the codebase because of differences between server/serverless (as the divergence is no longer needed)


So basically we're bringing optimizations found in optimizing sites hosted on ZEIT Now to all users including those who use the `server` target.

Also this prepares for the future:

- Will eventually allow for streaming rendering (this is one of the reasons Gerald is working on this)
- Will improve React Concurrent mode support


These updates will be made **without breaking changes** as usual, including custom server for example. We have an extensive integration test suite to ensure this.

**Reply by garnerp** [reply]
@timneutkens Thanks for the detailed explanation and the work that Zeit is leading.  Amazing!

**Reply by timneutkens** [reply]
Happy to explain! I'm personally very excited about these changes.
